### PR TITLE
feat: add support for ALB in azure

### PR DESCRIPTION
**Pull request by kedar-lanka**
- Support for ALB has been introduced as part this change
- As part of these changes, an Azure application gateway is created and based on users selection provision for creating a self signed certificate, which will be attached to ALB, has also been added.
- Along with ALB, Azure modules have been refactored similar to AWS
- `bicepconfig.json` has been added for future lint checks
- Also, `common.bicep` has been added for combining multiple common params across ALB, NLB & No-LB

### CODE REVIEW

File: deployments/azure/templates/bicep/all-in-one/all-in-one-with-alb.bicep [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+
+@description('password (if present) for certificate, ignored if UseSelfSignedCertForALB is enabled')
+@secure()
+param ApplicationgateWayCertificatePassword string = ''
+
+@description('Dns label prefix for ALB')
+param dnsLabelPrefix string
+
+@description('Join token for the Jupyter Labs service')
+param JupyterToken string
```
Comment (alicia-td) [inline deployments/azure/templates/bicep/all-in-one/all-in-one-with-alb.bicep:59]: For a better user experience, I think having a default for the JupyterToken would be helpful. 

On the develop's version, the default is set to `[uniqueString(subscription().id, utcNow())]`: https://github.com/Teradata/ai-unlimited/blob/7b52ba8c27c6b92c5249ac447c94819b606c584e/deployments/azure/bicep/all-in-one.bicep#L78

File: deployments/azure/templates/init/network.bicep [hunk 2]
```diff
@@ -1,29 +1,47 @@
...
+
   properties: {
     addressPrefix: subnetCidr
     privateEndpointNetworkPolicies: 'Enabled'
     privateLinkServiceNetworkPolicies: 'Enabled'
   }
 }
 
-output subnetName string = subnet.name
+resource gwSubnet 'Microsoft.Network/virtualNetworks/subnets@2022-11-01' = if (deployAlbSubnet) {
```
Comment (alicia-td) [inline deployments/azure/templates/init/network.bicep:33]: I'm using the resources arm template to deploy with, and I enabled deployAlbSubnet, and keep running in this issue:
Screenshot 2024-04-15 at 10 47 43 AM

It is consistent for me, and I have run with different `name` values. Are you able to reproduce this issue?

File: deployments/azure/templates/bicep/all-in-one/all-in-one-with-alb.bicep [hunk 3]
```diff
@@ -0,0 +1,328 @@
+targetScope = 'subscription'
```
Comment (alicia-td) [inline deployments/azure/templates/bicep/all-in-one/all-in-one-with-alb.bicep:1]: Very nice!

I just tested this all-in-one-with-alb arm template and verified it deploys successfully (using dev workspaces docker image with arm changes - `aliciadnguyen/regulus-workspaces:169a0d5`):

* This is with using `UseSelfSignedCertForALB = true`

Screenshot 2024-04-15 at 11 54 50 AM

Comment (alicia-td) [reply deployments/azure/templates/bicep/all-in-one/all-in-one-with-alb.bicep:1]: I tried passing in `ApplicationgateWayCertificate`, but I am running into this:
Screenshot 2024-04-15 at 12 15 58 PM

Screenshot 2024-04-15 at 12 16 18 PM

I used these commands to create the SSL certs and getting the base64 content of the cert: https://learn.microsoft.com/en-us/azure/application-gateway/self-signed-certificates

I believe I am not doing it correctly. Would you have the command for me to create the cert? Thank you.

File: deployments/azure/templates/bicep/all-in-one/all-in-one-with-nlb.bicep [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    cloudInitData: cloudInitData
+    usePersistentVolume: UsePersistentVolume
+    persistentVolumeSize: PersistentVolumeSize
+    existingPersistentVolume: ExistingPersistentVolume
+    nlbName: AiUnlimitedName
+    nlbPoolNames: nlb.outputs.nlbPools
+    usePublicIp: AllowPublicSSH
+    tags: Tags
+  }
+}
```
Comment (alicia-td) [inline deployments/azure/templates/bicep/all-in-one/all-in-one-with-nlb.bicep]: I tried deploying an NLB deployment, but am running into public IP address conflicts:
@
Screenshot 2024-04-15 at 12 56 31 PM

Is the nlb-inbound DNS label supposed to be the same as the ai-unlimited instance's DNS label?

File: deployments/azure/templates/bicep/all-in-one/all-in-one-without-lb.bicep [hunk 5]
```diff
@@ -34,13 +34,13 @@ param SecurityGroup string = 'AiUnlimitedSecurityGroup'
 param AccessCIDRs array = [ '0.0.0.0/0' ]
 
 @description('port to access the Jupyter Labs UI.')
-param JupyterHttpPort string = '8888'
+param JupyterHttpPort int = 8888
```
Comment (alicia-td) [inline deployments/azure/templates/bicep/all-in-one/all-in-one-without-lb.bicep:37]: Nice! I was able to deploy this template `all-in-one-without-lb` successfully:
Screenshot 2024-04-15 at 1 57 14 PM

**Review by alicia-td — COMMENTED** [review]
Changes look good!

I just left a couple of comments of some issues I ran into.

Also, I think it's fine to update the arm JSON templates in this PR and to create a JIRA to automate building the bicep -> ARM JSON templates.

File: deployments/azure/templates/init/network.bicep (hunk 2 above)
Comment (alicia-td) [reply deployments/azure/templates/init/network.bicep:33]: I got around this issue by declaring the subnet inside the network resource:
```
resource network 'Microsoft.Network/virtualNetworks@2022-11-01' = {
  name: networkName
  location: location

  properties: {
    addressSpace: {
      addressPrefixes: networkCidr
    }
    subnets: [
      {
        name: networkName
        properties: {
          addressPrefix: subnetCidr
          privateEndpointNetworkPolicies: 'Enabled'
          privateLinkServiceNetworkPolicies: 'Enabled'
        }
      }
    ]
  }
  tags: tags
}
```

File: deployments/azure/templates/bicep/all-in-one/all-in-one-with-nlb.bicep (hunk 4 above)
Comment (alicia-td) [reply deployments/azure/templates/bicep/all-in-one/all-in-one-with-nlb.bicep]: Was able to resolve this by changing their dns labels to be different and to not able public ip for instances with nlb

Comment (alicia-td) [reply deployments/azure/templates/bicep/all-in-one/all-in-one-with-nlb.bicep]: I was able to deploy NLB deployment with slight changes:
Screenshot 2024-04-16 at 3 29 53 PM

**Review by alicia-td — APPROVED** [review]

**Review by teralauritsen — APPROVED** [review]

**Pull request merged**